go/marshal/decode.go

```diff
@@ -0,0 +1,227 @@
+		return "Cannot unmarshal into Go non pointer of type " + e.Type.String()
+	}
+	return "Cannot unmarshal into Go nil pointer of type " + e.Type.String()
+}
+
+// UnmarshalTypeMismatchError describes a Noms value that was not appropriate for a value of a specific Go type.
+type UnmarshalTypeMismatchError struct {
+	Value   types.Value
+	Type    reflect.Type // type of Go value it could not be assigned to
+	details string
+}
+
+func (e *UnmarshalTypeMismatchError) Error() string {
+	return "Cannot unmarshal " + e.Value.Type().Describe() + " into Go value of type " + e.Type.String() + e.details
+}
+
+// OverflowError describes a Noms value that was not appropriate for a value of a specific Go number type.
+type OverflowError struct {
+	Value types.Number
+	Type  reflect.Type // type of Go value it could not be assigned to
+}
+
+func (e *OverflowError) Error() string {
+	return fmt.Sprintf("Cannot unmarshal %g into Go value of type %s", e.Value, e.Type)
+}
+
+type decoderFunc func(v types.Value, rv reflect.Value)
+
+func typeDecoder(t reflect.Type) decoderFunc {
+	switch t.Kind() {
+	case reflect.Bool:
+		return boolDecoder
+	case reflect.Float32, reflect.Float64:
+		return floatDecoder
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return intDecoder
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		return uintDecoder
+	case reflect.String:
+		return stringDecoder
+	case reflect.Struct, reflect.Interface:
+		return structDecoder(t)
+	default:
+		panic(&UnsupportedTypeError{Type: t})
+	}
+}
+func boolDecoder(v types.Value, rv reflect.Value) {
+	if b, ok := v.(types.Bool); ok {
+		rv.SetBool(bool(b))
+	} else {
+		panic(&UnmarshalTypeMismatchError{v, rv.Type(), ""})
+	}
+}
+
+func stringDecoder(v types.Value, rv reflect.Value) {
+	if s, ok := v.(types.String); ok {
+		rv.SetString(string(s))
+	} else {
+		panic(&UnmarshalTypeMismatchError{v, rv.Type(), ""})
+	}
+}
+
+func floatDecoder(v types.Value, rv reflect.Value) {
+	if n, ok := v.(types.Number); ok {
+		rv.SetFloat(float64(n))
+	} else {
+		panic(&UnmarshalTypeMismatchError{v, rv.Type(), ""})
+	}
+}
+
+func intDecoder(v types.Value, rv reflect.Value) {
+	if n, ok := v.(types.Number); ok {
+		i := int64(n)
+		if rv.OverflowInt(i) {
+			panic(&OverflowError{n, rv.Type()})
+		}
+		rv.SetInt(i)
+	} else {
+		panic(&UnmarshalTypeMismatchError{v, rv.Type(), ""})
+	}
+}
+
+func uintDecoder(v types.Value, rv reflect.Value) {
+	if n, ok := v.(types.Number); ok {
+		u := uint64(n)
+		if rv.OverflowUint(u) {
+			panic(&OverflowError{n, rv.Type()})
+		}
+		rv.SetUint(u)
+	} else {
+		panic(&UnmarshalTypeMismatchError{v, rv.Type(), ""})
+	}
+}
+
+var structDecoderCache struct {
+	sync.RWMutex
+	m map[reflect.Type]decoderFunc
+}
+
+type decField struct {
+	name    string
+	decoder decoderFunc
+	index   int
+}
+
+func structDecoder(t reflect.Type) decoderFunc {
+	// TODO: Is this handling wrong noms type?
+	if t.Implements(nomsValueType) {
```

Comment:
interfaces are types too but I agree that nomsValueInterface is better (more specific)